common/functions/src/scalars/strings/export_set.rs

```diff
@@ -0,0 +1,1659 @@
+    }
+
+    pub fn desc() -> FunctionDescription {
+        FunctionDescription::creator(Box::new(Self::try_create))
+            .features(FunctionFeatures::default().deterministic())
+    }
+}
+
+impl Function for ExportSetFunction {
+    fn name(&self) -> &str {
+        &*self.display_name
+    }
+
+    fn variadic_arguments(&self) -> Option<(usize, usize)> {
+        Some((3, 5))
+    }
+
+    fn nullable(&self, _input_schema: &DataSchema) -> Result<bool> {
+        Ok(true)
+    }
+
+    fn return_type(&self, args: &[DataType]) -> Result<DataType> {
+        if !args[0].is_integer() && args[0] != DataType::String && args[0] != DataType::Null {
+            return Err(ErrorCode::IllegalDataType(format!(
+                "Expected integer or string or null, but got {}",
+                args[0]
+            )));
+        }
+        if !args[1].is_integer() && args[1] != DataType::String && args[1] != DataType::Null {
+            return Err(ErrorCode::IllegalDataType(format!(
+                "Expected integer or string or null, but got {}",
+                args[1]
+            )));
+        }
+        if !args[2].is_integer() && args[2] != DataType::String && args[2] != DataType::Null {
+            return Err(ErrorCode::IllegalDataType(format!(
+                "Expected integer or string or null, but got {}",
+                args[2]
+            )));
+        }
+        if args.len() > 3
+            && !args[3].is_integer()
+            && args[3] != DataType::String
+            && args[3] != DataType::Null
+        {
+            return Err(ErrorCode::IllegalDataType(format!(
+                "Expected integer or string or null, but got {}",
+                args[3]
+            )));
+        }
+        if args.len() > 4
+            && !args[4].is_integer()
+            && args[4] != DataType::String
+            && args[4] != DataType::Null
+        {
+            return Err(ErrorCode::IllegalDataType(format!(
+                "Expected integer or string or null, but got {}",
+                args[4]
+            )));
+        }
+
+        Ok(DataType::String)
+    }
+
+    fn eval(&self, columns: &DataColumnsWithField, r: usize) -> Result<DataColumn> {
+        let b_column = columns[0].column().cast_with_type(&DataType::UInt64)?;
+        let e_column = columns[1].column().cast_with_type(&DataType::String)?;
+        let d_column = columns[2].column().cast_with_type(&DataType::String)?;
+        let s_column = if columns.len() > 3 {
+            columns[3].column().cast_with_type(&DataType::String)?
+        } else {
+            DataColumn::Constant(DataValue::String(Some(vec![44])), r)
+        };
+        let n_column = if columns.len() > 4 {
+            columns[4].column().cast_with_type(&DataType::UInt64)?
+        } else {
+            DataColumn::Constant(DataValue::UInt64(Some(64)), r)
+        };
+
+        let r_column: DataColumn = match (b_column, e_column, d_column, s_column, n_column) {
+            // #00000
```

Comment:
What a looooooooooooooong match.

Reply:
Hello, @mshauneu  really thanks for your effort to add this function.

I tried to review this function myself and I found it's hard to review and maintain/refactor the codes,  left some comments to improve this.

1. We need to force  `b_column`, `n_column` to be constant datavalue column because it covers most cases, returns errors if not.

2.  Then better to force `d_column` to be a constant datavalue column, it's unsound to have the delimiter be dynamic, returns errors if not.

Then we just cut down 2^5 = 32 branches to be  2 ^ 2 = 4 branches.

Reply:
I was trying to follow mysql:
```
SELECT EXPORT_SET(number, number, number, number, number)  FROM numbers WHERE number < 5  ORDER BY number;
+----------------------------------------------------+
| EXPORT_SET(number, number, number, number, number) |
+----------------------------------------------------+
|                                                    |
| 1                                                  |
| 222                                                |
| 33333                                              |
| 4444444                                            |
+----------------------------------------------------+
```
Anyway It is easy to shrink than expand, so please take a look to last commit.

Reply:
>  SELECT EXPORT_SET(number, number, number, number, number)  FROM numbers WHERE number < 5  ORDER BY number;

In this case, we can add another branch, which converts all columns into Series using `to_array`. It may be inefficient but cover all cases.